query/src/main/scala/filodb/query/exec/RemoteExec.scala

```diff
@@ -95,6 +100,83 @@ trait RemoteExec extends LeafExecPlan with StrictLogging {
   }
 }
 
+trait GrpcChannelManager {
+
+  /**
+   * Based on the endpointURL borrow the Grpc channel to use
+   *
+   * @param endpointUrl
+   * @return io.grpc.Channel
+   */
+  def borrowChannel(endpointUrl: String): ManagedChannel
+
+  /**
+   * Returns the channel to the manager, the implementation may choose to close the channel of retain for later reuse
+   * @param channel
+   */
+  def returnChannel(channel: ManagedChannel): Unit
+
+  def shutdown(): Unit
+
+}
+
+abstract class BaseChannelManager extends GrpcChannelManager {
+
+  def buildChannelFromEndpoint(endpointUrl: String): ManagedChannel = {
+
+    val grpcConfig = GlobalConfig.defaultFiloConfig.getConfig("grpc")
+    val idleTimeout = grpcConfig.getInt("idle-timeout-seconds")
+    val keepAliveTime = grpcConfig.getInt("keep-alive-time-seconds")
+    val keepAliveTimeOut = grpcConfig.getInt("keep-alive-timeout-seconds")
+    NettyChannelBuilder
+      .forTarget(endpointUrl)
+      // TODO: Configure this to SSL/Plain text later based on config, currently only Plaintext supported
+      .negotiationType(NegotiationType.PLAINTEXT)
+      .idleTimeout(idleTimeout, TimeUnit.SECONDS)
+      .keepAliveTime(keepAliveTime, TimeUnit.SECONDS)
+      .keepAliveTimeout(keepAliveTimeOut, TimeUnit.SECONDS)
+      .keepAliveWithoutCalls(true)
+      .build()
+  }
+}
+
+/**
+ * Simple but inefficient implementation that does not reuse channels. Simply builds the channel, possibly multiple
+ * to same endpoint and shut it down when not needed
+ */
+class SimpleGrpcChannelManager extends BaseChannelManager {
+
+  override def borrowChannel(endpointUrl: String): ManagedChannel = buildChannelFromEndpoint(endpointUrl)
+
+  override def returnChannel(channel: ManagedChannel): Unit = channel.shutdown()
+
+  override def shutdown(): Unit = {}
+}
+
+// TODO: Test for concurrency and performance. The handling for retrying after connection loss needs to be worked on
+class ReusableGRPCChannelManager extends BaseChannelManager {
```

Comment:
who is responsible for keep-alive (or termination) of unused connections?

Reply:
Thanks @vishramachandran this class is still not completely developed. I need to put in some work here. The simple implementation is what I used for current tests locally, we should use this implementation to manage channels. I will be working on this more in coming days.